### DISCUSSION TITLE
Dynamically create model from base model excluding specific fields

### DISCUSSION THREAD

**Discussion by jordantshaw** [pydantic/pydantic#6863] [Question]
Is it possible to use the `create_model` function to dynamically create a model from a given base model, excluding specific fields. Below is an example of what I am trying to do.

```py
from pydantic import BaseModel, create_model

class User(BaseModel):
    name: Optional[str] = None
    id: int 
    nickname: Optional[str] = None


UserDisplay = create_model('UserDisplay', __base__=User, exclude_fields=['name'])

```

**Comment by braindevices** [comment]
Your code does not make any sense at all. Next time please provide working code.

- there is no BaseClass
- your name and nickname member default value does not match your type annotation.
- you did not provide any code to show expected result.

**Comment by braindevices** [comment]
I have to guess what you actually want:
```py
from typing import Optional
from pydantic import BaseModel, create_model, Field

class User(BaseModel):
    name: Optional[str] = None
    id: int 
    nickname: Optional[str] = None


UserDisplay = create_model('UserDisplay', __base__=User, name = (str, Field(None, exclude=True)))
UserDisplay(name="test", id=1).model_dump()
```

Output is `{'id': 1, 'nickname': None}`

**Reply by jordantshaw** [reply]
Thats not quite what Im looking for.  Instead I would like that field to be ignored during model creation, so do not include that field in the new model, even though its in the base.

**Comment by jadshaker** [comment]
In my case, i didn't care if the functions were carried on or not, so this the approach I used
```py
NewModel = create_model(
            f"{model.__name__}New",
            **{k: (v.annotation, None) for k, v in MyOldModel.model_fields.items() if k != "id"}
        )
```
In my case I wanted to exclude the "id" from the model

**Comment by v-anzhurov** [comment]
Any update on that one ?

I'd like to `create_model` from `__base__` excluding fields from base model but keeping all then config and validators. In other words, I'd like to create reduced model (in terms of fields) from extended one

**Comment by xamcost** [comment]
This looks like an [XY problem](https://xyproblem.info/). The real problem seems to be how to model an entity in such a way that you can validate it when users create an instance with a subset of its attributes (for instance via a post request) **and** create a Python instance of the full entity (for instance after retrieving it from a database). For this, a common approach is to do 2 to 3 models:
- an abstract base class containing all the minimal set of fields of the entity
- a class to validate created instance
- a class to instantiate the entity and use it your system

For example:
```python
from pydantic import BaseModel, Field


class EntityBase(BaseModel):
    name: str = Field(..., description="Name of the entity")
    description: str = Field(default="", description="Description of the entity")


class EntityCreate(EntityBase):
    pass


class Entity(EntityBase):
    id: str = Field(..., description="Unique identifier for the entity")
```

This approach makes the intent of the model clear, while keeping a good separation of concerns.

**Reply by jordantshaw** [reply]
The issue with this approach is its not dynamic. It would be awesome if you could create a sub class from the base class dynamically excluding a list of specific fields that could change depending on user input.

**Comment by StoneSteel27** [comment]
I have built an module to handle this issue, as this was crucial for developing removable parts of an context history of an AI agent, That I was building : https://github.com/StoneSteel27/pydantic-pick

Here is an example:
```py
from pydantic import BaseModel
from pydantic_pick import create_subset

class Profile(BaseModel):
    avatar_url: str
    billing_secret: str  # We want to drop this

class DBUser(BaseModel):
    id: int
    username: str
    password_hash: str  # And drop this
    profiles: list[Profile]

    def check_password(self, guess: str) -> bool:
        # This method relies on password_hash
        return self.password_hash == guess

# Create a subset using dot-notation to drill into nested lists
PublicUser = create_subset(
    DBUser, 
    ("id", "username", "profiles.avatar_url"), 
    "PublicUser"
)

user = PublicUser(id=1, username="alice", profiles=[{"avatar_url": "img.png"}])

# Because password_hash was omitted, AST parsing automatically drops check_password
# Calling user.check_password("secret") will raise a custom AttributeError 
# explaining it was intentionally omitted during extraction
```

It can handle most types like Union, Enums, List, Dict, Set, Tuple, Literals, Annotated. And also handles Field, constants, ClassVars.
It will take care of class based wrappers, function based wrappers with AST parsing. this means, `@computed_fields` and validators are handled

If you have any issues, please feel free to raise in my repo
